src/main/java/com/redhat/rhjmc/containerjfr/net/web/http/api/v2/CertificatePostHandler.java

```diff
@@ -0,0 +1,172 @@
+
+    private final Environment env;
+    private final FileSystem fs;
+    private final Logger logger;
+
+    private static final String TRUSTSTORE_DIR = "TRUSTSTORE_DIR";
+
+    private Function<File, FileOutputStream> outputStreamFunction;
+    private CertificateValidator certValidator;
+
+    @Inject
+    CertificatePostHandler(
+            AuthManager auth,
+            Environment env,
+            FileSystem fs,
+            Logger logger,
+            @Named("OutputStreamFunction") Function<File, FileOutputStream> outputStreamFunction,
+            CertificateValidator certValidator) {
+        super(auth);
+        this.env = env;
+        this.fs = fs;
+        this.logger = logger;
+        this.outputStreamFunction = outputStreamFunction;
+        this.certValidator = certValidator;
+    }
+
+    @Override
+    public ApiVersion apiVersion() {
+        return ApiVersion.V2;
+    }
+
+    @Override
+    public HttpMethod httpMethod() {
+        return HttpMethod.POST;
+    }
+
+    @Override
+    public String path() {
+        return basePath() + PATH;
+    }
+
+    @Override
+    public boolean isAsync() {
+        return false;
+    }
+
+    @Override
+    public boolean isOrdered() {
+        return true;
+    }
+
+    @Override
+    @SuppressFBWarnings("RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE")
+    public void handleAuthenticated(RoutingContext ctx) throws Exception {
+        FileUpload cert = null;
+        for (FileUpload fu : ctx.fileUploads()) {
+            if ("cert".equals(fu.name())) {
+                cert = fu;
+                break;
+            }
+        }
+
+        if (cert == null) {
+            throw new HttpStatusException(
+                    400, "A file named \"cert\" was not included in the request");
+        }
+
+        Path certPath = fs.pathOf(cert.uploadedFileName());
+
+        if (!env.hasEnv(TRUSTSTORE_DIR)) {
+            throw new HttpStatusException(500, "Truststore directory not set");
+        }
+
+        String truststoreDir = env.getEnv(TRUSTSTORE_DIR);
+        String filePath = fs.pathOf(truststoreDir, cert.fileName()).normalize().toString();
+
+        if (fs.exists(fs.pathOf(filePath))) {
+            throw new HttpStatusException(409, filePath + " Certificate already exists");
+        }
+
+        File certFile = new File(filePath);
+
+        try (InputStream fis = fs.newInputStream(certPath);
+                DataInputStream dis = new DataInputStream(fis);
+                FileOutputStream out = outputStreamFunction.apply(certFile)) {
+
+            byte[] bytes = new byte[dis.available()];
+            dis.readFully(bytes);
+            ByteArrayInputStream bytestream = new ByteArrayInputStream(bytes);
+            Certificate certificate = certValidator.verify(bytestream);
+            byte[] buf = certificate.getEncoded();
+
+            out.write(buf);
+        } catch (Exception e) {
+            throw new HttpStatusException(500, e.getMessage());
```

Comment:
This `catch` clause should probably be removed - the parent `AbstractAuthenticatedRequestHandler` class already has handling for exceptions thrown by subclasses, including general Exceptions which get wrapped into `500`s similar to what has been repeated here. IMO it would be better to allow the abstract parent class to handle this so that the behaviour across all concrete implementations is uniform.

Reply:
Yes, that is definitely better!